Added Support for domain based group members on Nano Server

AndreasSobczyk opened the pull request:
Resolution for Issue [38](https://github.com/PowerShell/PSDscResources/issues/38)

Adds support for domain based group members on nano server.
This is a very important feature to be able to control the local administrators on a nano server without having GPOs/Restricted Groups.

If i missed anything i would gladly change :)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/psdscresources/39)
<!-- Reviewable:end -->

msftclas commented:
Hi __@AndreasSobczyk__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftgits commented:
Hi, I am closing and re-opening this PR to bump the CLA bot. Sorry for the inconvenience!

msftclas commented:
Hi __@AndreasSobczyk__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, MSBOT;

kwirkykat commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[README.md, line 508 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCm_iEIe_E5FDCWuDm:-KdCm_iEIe_E5FDCWuDn:b8cdqeo) ([raw file](https://github.com/powershell/psdscresources/blob/5c7fe19f6102bf1411b9604ef507d95c46e18a26/README.md#L508)):*
> ```Markdown
> * Enable codecov.io code coverage reporting
> - Group
>     - Added support for domain based group members on Nano server.
> ```

This change does not make domain based members work on Nano server for Group. Only the local members can be retrieved and set still. Please update the release note

---

*[DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1, line 968 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCmiji7N5Zpoea-4tw:-KdCmiji7N5Zpoea-4tx:b1qh2u8) ([raw file](https://github.com/powershell/psdscresources/blob/5c7fe19f6102bf1411b9604ef507d95c46e18a26/DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1#L968)):*
> ```PowerShell
>             elseif ($PSBoundParameters.ContainsKey('MembersToInclude') -or $PSBoundParameters.ContainsKey('MembersToExclude'))
>             {
>                 [array]$groupMembers = Get-MembersOnNanoServer -Group $group
> ```

This cast and the one below are the actual fixes for the bug described.
There are no changes really needed in Get-MembersOnNanoServer.

It's a (extremely confusing) behavior of PowerShell that even if we declare that the function returns an array, it will return only a single object outside of an array if an array with only one object is returned.

---

*[DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1, line 1501 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCm84AlccKsVk8Tq0B:-KdCm84AlccKsVk8Tq0C:b-wcr5k8) ([raw file](https://github.com/powershell/psdscresources/blob/5c7fe19f6102bf1411b9604ef507d95c46e18a26/DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1#L1501)):*
> ```PowerShell
> function Get-MembersOnNanoServer
> {
>     [OutputType([System.Array[]])]
> ```

Why are you changing the output of this function from an array of strings to an array of arrays of unknown types?

---

*[DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1, line 1525 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCmQt6MbzkYqnx5LR0:-KdCmQt6MbzkYqnx5LR1:b-6vmz0) ([raw file](https://github.com/powershell/psdscresources/blob/669a19b1664a0b63807135f967756130e173a51d/DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1#L1525)):*
> ```PowerShell
>         else
>         {
>             Write-Verbose -Message ($script:localizedData.MemberIsNotALocalUser -f $groupMember.Name,
> ```

Why are you removing the verbose message here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/psdscresources/39)*
<!-- Sent from Reviewable.io -->

AndreasSobczyk commented:
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[README.md, line 508 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCm_iEIe_E5FDCWuDm:-KdPdkVsEx1IMqn_Q82F:bgngz3o) ([raw file](https://github.com/powershell/psdscresources/blob/5c7fe19f6102bf1411b9604ef507d95c46e18a26/README.md#L508)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

This change does not make domain based members work on Nano server for Group. Only the local members can be retrieved and set still. Please update the release note
</blockquote></details>

The changes at line 1526 and 1527 added the support for domain accounts.
I tested it to replace the Restricted Groups from GPOs.

---

*[DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1, line 1501 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCm84AlccKsVk8Tq0B:-KdPbzJkWYTil9_iSge9:b-13ufwz) ([raw file](https://github.com/powershell/psdscresources/blob/5c7fe19f6102bf1411b9604ef507d95c46e18a26/DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1#L1501)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

Why are you changing the output of this function from an array of strings to an array of arrays of unknown types?
</blockquote></details>

Ahh I thought i actually removed this again, was testing if i could force it to return an array. I have changed this back to string.

---

*[DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1, line 1525 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCmQt6MbzkYqnx5LR0:-KdPcmPcC1Ae-3CG29GH:b-lwao31) ([raw file](https://github.com/powershell/psdscresources/blob/669a19b1664a0b63807135f967756130e173a51d/DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1#L1525)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

Why are you removing the verbose message here?
</blockquote></details>

Looked like the verbose part was only to notice that these account wouldn't be added  but with the change to Get-MembersOnNanoServer it will include domain accounts.
I added it back into the code as you can never get to much verbose :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/psdscresources/39)*
<!-- Sent from Reviewable.io -->

AndreasSobczyk commented:
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1, line 968 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCmiji7N5Zpoea-4tw:-KdPeut1MxRvsGQJ--zK:b3as9x7) ([raw file](https://github.com/powershell/psdscresources/blob/5c7fe19f6102bf1411b9604ef507d95c46e18a26/DscResources/MSFT_GroupResource/MSFT_GroupResource.psm1#L968)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary><blockquote>

This cast and the one below are the actual fixes for the bug described.
There are no changes really needed in Get-MembersOnNanoServer.

It's a (extremely confusing) behavior of PowerShell that even if we declare that the function returns an array, it will return only a single object outside of an array if an array with only one object is returned.
</blockquote></details>

Yes this fixes the bug. The other changes at line 1526 and 1527 adds support for domain accounts.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/psdscresources/39)*
<!-- Sent from Reviewable.io -->

kwirkykat commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[README.md, line 508 at r1](https://reviewable.io:443/reviews/powershell/psdscresources/39#-KdCm_iEIe_E5FDCWuDm:-KdbCWtNQ8Xx74lXQGOm:b-199dkg) ([raw file](https://github.com/powershell/psdscresources/blob/5c7fe19f6102bf1411b9604ef507d95c46e18a26/README.md#L508)):*
<details><summary><i>Previously, AndreasSobczyk (AndreasSobczyk) wrote…</i></summary><blockquote>

The changes at line 1526 and 1527 added the support for domain accounts.
I tested it to replace the Restricted Groups from GPOs.
</blockquote></details>

Sorry. I was confused because the old full SKU code for group is so complicated for domain users. I asked around and this should make domain users just work on Nano :+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/psdscresources/39#-:-KdbDyDl0LRzcncEOfqo:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kwirkykat commented:
Merging this without the patch code coverage test passing since this code would only be tested on Nano server currently, which has not been set up to run with the automated code coverage test.
